sdk/program/src/bignum.rs

```diff
@@ -0,0 +1,784 @@
+//! @brief Solana Rust-based BigNumber for bpf-programs
+
+use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
+use std::fmt;
+
+#[repr(C)]
+#[derive(
+    BorshSerialize,
+    BorshDeserialize,
+    BorshSchema,
+    Clone,
+    Eq,
+    PartialEq,
+    Ord,
+    PartialOrd,
+    Hash,
+    AbiExample,
+)]
+pub struct BigNumber {
+    value: Vec<u8>,
+    negative: bool,
+}
+
+// pub struct BigNumber(u64);
+
+impl BigNumber {
+    /// Returns a BigNumber with initial value of 0
+    pub fn new() -> Self {
+        Self {
+            negative: false,
+            value: vec![0u8],
+        }
+    }
+
+    /// Returns the size, in bytes, of BigNum. Typically used in
+    /// assessing buffer size needed to fetch the bytes for serialization, etc.
+    pub fn size_in_bytes(&self) -> usize {
+        self.value.len()
+    }
+
+    /// Returns a BigNumber with initial value set to a u32 value
+    pub fn from_u32(val: u32) -> Self {
+        if val == 0 {
+            Self {
+                negative: false,
+                value: vec![0u8],
+            }
+        } else {
+            #[cfg(not(target_arch = "bpf"))]
+            {
+                use openssl::bn::BigNum;
+                let value = BigNum::from_u32(val).unwrap().to_vec();
+                Self {
+                    negative: false,
+                    value,
+                }
+            }
+            #[cfg(target_arch = "bpf")]
+            {
+                extern "C" {
+                    fn sol_bignum_from_u32(bytes_ptr: *mut u8, bytes_len: u64, val: u64) -> u64;
+                }
+                let vec_size = {
+                    match val {
+                        0..=255 => 1,
+                        256..=65_535 => 2,
+                        65_536..=16_777_215 => 3,
+                        _ => 4,
+                    }
+                };
+                let mut value = Vec::<u8>::with_capacity(vec_size);
+                unsafe {
+                    sol_bignum_from_u32(
+                        value.as_mut_ptr() as *mut _ as *mut u8,
+                        vec_size as u64,
+                        val as u64,
+                    );
+                    value.set_len(vec_size);
```

Comment:
Why do we have to set the length here?

Reply:
I allocate with capacity prior to the call and populate from 0..capacity in the syscall but unless the length is set after, reading from the vector fails as it is not bounded.